File: src/main/java/com/redbox/domain/user/service/UserService.java
package com.redbox.domain.user.service;

import com.redbox.domain.auth.dto.CustomUserDetails;
import com.redbox.domain.donation.repository.DonationGroupRepository;
import com.redbox.domain.request.dto.ListResponse;
import com.redbox.domain.request.dto.RequestFilter;
import com.redbox.domain.request.entity.Request;
import com.redbox.domain.request.repository.RequestRepository;
import com.redbox.domain.user.dto.*;
import com.redbox.domain.user.entity.User;
import com.redbox.domain.user.exception.DuplicateEmailException;
import com.redbox.domain.user.exception.EmailNotVerifiedException;
import com.redbox.domain.user.exception.PasswordNotMatchException;
import com.redbox.domain.user.exception.UserNotFoundException;
import com.redbox.domain.user.exception.PasswordMismatchException;
import com.redbox.domain.user.repository.EmailVerificationCodeRepository;
import com.redbox.domain.user.repository.UserRepository;
import com.redbox.global.entity.PageResponse;
import com.redbox.global.util.RandomCodeGenerator;
import com.redbox.global.util.email.EmailSender;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.thymeleaf.context.Context;
import org.thymeleaf.spring6.SpringTemplateEngine;

@Service
@RequiredArgsConstructor
public class UserService {

    private final SpringTemplateEngine templateEngine;
    private final EmailSender emailSender;
    private final EmailVerificationCodeRepository emailVerificationCodeRepository;
    private final PasswordEncoder passwordEncoder;
    private final UserRepository userRepository;
    private final DonationGroupRepository donationGroupRepository;
    private final RequestRepository requestRepository;

    // 현재 로그인한 사용자의 전체 정보 조회
    public User getCurrentUser() {
        CustomUserDetails userDetails = getCustomUserDetails();
        return userRepository.findByEmail(userDetails.getUsername())
                .orElseThrow(UserNotFoundException::new);
    }

    // 현재 로그인한 user_id
    public Long getCurrentUserId() {
        return getCustomUserDetails().getUserId();
    }

    private CustomUserDetails getCustomUserDetails() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        return (CustomUserDetails) authentication.getPrincipal();
    }

    private String createEmailContent(String templateName, String variableName, String variableValue) {
        Context context = new Context();
        context.setVariable(variableName, variableValue);
        return templateEngine.process(templateName, context);
    }

    public void sendVerificationCode(VerificationCodeRequest request) {
        // 이미 회원가입이 된 이메일인지 확인
        if (isDuplicatedEmail(request.getEmail())) {
            throw new DuplicateEmailException();
        }
        String verificationCode = RandomCodeGenerator.generateRandomCode();
        String subject = "[Redbox] 이메일 인증 코드입니다.";
        String content = createEmailContent("verification-code-email", "verificationCode", verificationCode);
        emailSender.sendMail(request.getEmail(), subject, content);
        emailVerificationCodeRepository.save(request.getEmail(), verificationCode);
    }

    private boolean isDuplicatedEmail(String email) {
        return userRepository.existsByEmail(email);
    }

    @Transactional
    public boolean validateVerificationCode(ValidateVerificationCodeRequest request) {
        return emailVerificationCodeRepository.getVerificationCodeByEmail(request.getEmail())
                .filter(verificationCode -> verificationCode.equals(request.getVerificationCode()))
                .map(verificationCode -> {
                    emailVerificationCodeRepository.deleteByEmail(request.getEmail());
                    return Boolean.TRUE;
                })
                .orElse(Boolean.FALSE);
    }

    @Transactional
    public SignupResponse signup(SignupRequest request) {
        // 이메일 인증이 완료되었는지 확인
        if (!request.isVerified()) {
            throw new EmailNotVerifiedException();
        }

        // 이미 회원가입이 된 이메일인지 확인
        if (isDuplicatedEmail(request.getEmail())) {
            throw new DuplicateEmailException();
        }

        String encodedPassword = encodePassword(request.getPassword());
        User user = SignupRequest.toEntity(request, encodedPassword);
        // 처음 회원가입 시 인증된 상태가 아니므로 직접 설정
        user.setCreatedBy(request.getEmail());
        user.setUpdatedBy(request.getEmail());

        userRepository.save(user);
        return new SignupResponse(user.getEmail(), user.getName());
    }

    private String encodePassword(String password) {
        return passwordEncoder.encode(password);
    }

    @Transactional
    public void resetPassword(ResetPasswordRequest request) {
        // 사용자 조회
        User user = userRepository.findByEmailAndName(request.getEmail(), request.getUsername())
                .orElseThrow(UserNotFoundException::new);

        // 임시 비밀번호 생성
        String tempPassword = RandomCodeGenerator.generateRandomCode();
        String encodedPassword = encodePassword(tempPassword);

        // 비밀번호 변경
        user.changePassword(encodedPassword);
        userRepository.save(user);

        // 이메일 전송
        String subject = "[Redbox] 임시 비밀번호 안내";
        String content = createEmailContent("temp-password-email", "tempPassword", tempPassword);
        emailSender.sendMail(request.getEmail(), subject, content);
    }

    @Transactional
    public FindIdResponse findUserId(FindIdRequest request) {
        String name = request.getUserName();
        String phoneNumber = request.getPhoneNumber();

        // 해당 정보로 사용자를 찾고, 없으면 예외 던짐
        String email = userRepository.findByNameAndPhoneNumber(name, phoneNumber)
                .orElseThrow(UserNotFoundException::new)
                .getEmail();

        return new FindIdResponse(email);
    }
  
    public UserInfoResponse getUserInfo() {
        User user = getCurrentUser();
        return new UserInfoResponse(user);
    }

    @Transactional
    public UserInfoResponse updateUserInfo(UpdateUserInfoRequest updateRequest) {
        User user = getCurrentUser();

        if (updateRequest.getName() != null) {
            user.changeName(updateRequest.getName());
        }
        if (updateRequest.getPhoneNumber() != null) {
            user.changePhoneNumber(updateRequest.getPhoneNumber());
        }
        if (updateRequest.getRoadAddress() != null) {
            user.changeRoadAddress(updateRequest.getRoadAddress());
        }
        if (updateRequest.getExtraAddress() != null) {
            user.changeExtraAddress(updateRequest.getExtraAddress());
        }
        if (updateRequest.getDetailAddress() != null) {
            user.changeDetailAddress(updateRequest.getDetailAddress());
        }
      
        userRepository.save(user);
        return new UserInfoResponse(user);
    }

    // 비밀 번호 일치 여부 확인 로직
    @Transactional
    public void dropUser(DropInfoRequest request) {

        // 현재 로그인한 사용자 조회
        User currentUser = getCurrentUser();

        // 입력받은 비밀번호와 현재 사용자의 비밀번호 비교
        if (!passwordEncoder.matches(request.getPassword(), currentUser.getPassword())) {
            throw new PasswordMismatchException();
        }

        currentUser.inactive();
        userRepository.save(currentUser);
    }
  
    @Transactional
    public void changePassword(UpdatePasswordRequest request) {
        if (!request.getPassword().equals(request.getPasswordConfirm())) {
            throw new PasswordNotMatchException();
        }

        User user = getCurrentUser();
        user.changePassword(encodePassword(request.getPassword()));
    }

    public CheckUserResponse checkUser(CheckUserRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                                         .orElseThrow(UserNotFoundException::new);

        return new CheckUserResponse(user.getId(), user.getName());
    }
  
    public PageResponse<DonationResponse> getDonations(int page, int size) {
        Pageable pageable = PageRequest.of(page - 1, size);
        return new PageResponse<>(donationGroupRepository.findAllWithReceiverNameByDonorId(getCurrentUserId(), pageable));
    }

    public PageResponse<ReceptionResponse> getReceptions(int page, int size) {
        Pageable pageable = PageRequest.of(page - 1, size);
        return new PageResponse<>(donationGroupRepository.findAllWithDonorNameByReceiverId(getCurrentUserId(), pageable));
    }

    public PageResponse<ListResponse> getRequests(int page, int size) {
        Pageable pageable = PageRequest.of(page -1, size, Sort.by("createdAt").descending());

        Page<Request> boardPage = requestRepository.findAllByUserIdAndNotDropStatus(getCurrentUserId(), pageable);
        Page<ListResponse> responsePage = boardPage.map(ListResponse::new);

        return new PageResponse<>(responsePage);
    }
}

File: migration/src/main/kotlin/com/redbox/domain/user/service/UserService.kt
package com.redbox.domain.user.service

import com.redbox.domain.community.funding.exception.UserNotFoundException
import com.redbox.domain.donation.dto.DonationListResponse
import com.redbox.domain.donation.facade.DonationFacade
import com.redbox.domain.redcard.dto.RedcardResponse
import com.redbox.domain.redcard.dto.RegisterRedcardRequest
import com.redbox.domain.redcard.dto.UpdateRedcardStatusRequest
import com.redbox.domain.redcard.facade.RedcardFacade
import com.redbox.domain.user.dto.*
import com.redbox.domain.user.entity.User
import com.redbox.domain.user.exception.DuplicateEmailException
import com.redbox.domain.user.exception.EmailNotVerifiedException
import com.redbox.domain.user.exception.PasswordMismatchException
import com.redbox.domain.user.exception.PasswordNotMatchException
import com.redbox.domain.user.repository.EmailVerificationCodeRepository
import com.redbox.domain.user.repository.UserRepository
import com.redbox.global.auth.service.AuthenticationService
import com.redbox.global.entity.PageResponse
import com.redbox.global.util.RandomCodeGenerator
import com.redbox.global.util.email.EmailSender
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.thymeleaf.context.Context
import org.thymeleaf.spring6.SpringTemplateEngine


@Service
class UserService(
    private val templateEngine: SpringTemplateEngine,
    private val emailSender: EmailSender,
    private val emailVerificationCodeRepository: EmailVerificationCodeRepository,
    private val passwordEncoder: PasswordEncoder,
    private val userRepository: UserRepository,
    private val redcardFacade: RedcardFacade,
    private val authenticationService: AuthenticationService,
    private val donationFacade: DonationFacade
    //private val donationGroupRepository: DonationGroupRepository,
    //private val fundingRepository: FundingRepository,

) {
    private fun isDuplicatedEmail(email: String): Boolean {
        return userRepository.existsByEmail(email)
    }

    private fun createEmailContent(templateName: String, variableName: String, variableValue: String): String {
        val context: Context = Context()
        context.setVariable(variableName, variableValue)
        return templateEngine.process(templateName, context)
    }

    private fun encodePassword(password: String): String = passwordEncoder.encode(password)

    fun sendVerificationCode(request: VerificationCodeRequest) {
        // 이미 회원가입이 된 이메일인지 확인
        if (isDuplicatedEmail(request.email)) {
            throw DuplicateEmailException()
        }
        val verificationCode = RandomCodeGenerator.generateRandomCode()
        val subject = "[Redbox] 이메일 인증 코드입니다."
        val content: String = createEmailContent("verification-code-email", "verificationCode", verificationCode)
        emailSender.sendMail(request.email, subject, content)
        emailVerificationCodeRepository.save(request.email, verificationCode)
    }

    @Transactional
    fun validateVerificationCode(request: ValidateVerificationCodeRequest): Boolean {
        val verificationCode = emailVerificationCodeRepository.getVerificationCodeByEmail(request.email)

        return if (verificationCode == request.verificationCode) {
            emailVerificationCodeRepository.deleteByEmail(request.email)
            true
        } else {
            false
        }
    }

    @Transactional
    fun signup(request: SignupRequest): SignupResponse {
        // 이메일 인증이 완료되었는지 확인
        if (!request.isVerified()) {
            throw EmailNotVerifiedException()
        }

        // 이미 회원가입이 된 이메일인지 확인
        if (isDuplicatedEmail(request.email)) {
            throw DuplicateEmailException()
        }

        val encodedPassword = encodePassword(request.password)
        val user = SignupRequest.toEntity(request, encodedPassword)
        // 처음 회원가입 시 인증된 상태가 아니므로 직접 설정
        user.setCreatedBy(request.email)
        user.setUpdatedBy(request.email)

        userRepository.save(user)
        return SignupResponse(user.email, user.name)
    }

    @Transactional
    fun resetPassword(request: ResetPasswordRequest) {
        // 사용자 조회
        val user = userRepository.findByEmailAndName(request.email, request.username) ?: throw UserNotFoundException()

        // 임시 비밀번호 생성
        val tempPassword = RandomCodeGenerator.generateRandomCode()
        val encodedPassword = encodePassword(tempPassword)

        // 비밀번호 변경
        user.changePassword(encodedPassword)
        userRepository.save(user)

        // 이메일 전송
        val subject = "[Redbox] 임시 비밀번호 안내"
        val content = createEmailContent("temp-password-email", "tempPassword", tempPassword)
        emailSender.sendMail(request.email, subject, content)
    }

    fun findUserId(request: FindIdRequest): FindIdResponse {
        val name = request.userName
        val phoneNumber = request.phoneNumber

        val email = userRepository.findByNameAndPhoneNumber(name, phoneNumber)?.email ?: throw UserNotFoundException()

        return FindIdResponse(email)
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    fun registerRedCard(request: RegisterRedcardRequest) {
        redcardFacade.registerRedCard(request)
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    @Transactional
    fun dropUser(request: DropInfoRequest) {
        // 현재 로그인한 사용자 조회
        val currentUser: User = authenticationService.getCurrentUser()

        // 입력받은 비밀번호와 현재 사용자의 비밀번호 비교
        if (!passwordEncoder.matches(request.password, currentUser.password)) {
            throw PasswordMismatchException()
        }

        currentUser.inactive()
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    fun getUserInfo(): UserInfoResponse {
        val user: User = authenticationService.getCurrentUser()
        return UserInfoResponse(user)
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    @Transactional
    fun updateUserInfo(request: UpdateUserInfoRequest): UserInfoResponse {
        val user: User = authenticationService.getCurrentUser()
        user.changeName(request.name)
        user.changePhoneNumber(request.phoneNumber)
        user.changeRoadAddress(request.roadAddress)
        user.changeExtraAddress(request.extraAddress)
        user.changeDetailAddress(request.detailAddress)
        userRepository.save(user)
        return UserInfoResponse(user)
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    @Transactional
    fun changePassword(request: UpdatePasswordRequest) {
        if (request.password != request.passwordConfirm) {
            throw PasswordNotMatchException()
        }

        val user: User = authenticationService.getCurrentUser()
        user.changePassword(encodePassword(request.password))
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    fun getRedcards(page: Int, size: Int): PageResponse<RedcardResponse> {
        return redcardFacade.getRedcards(page, size)
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    fun updateRedcardStatus(request: UpdateRedcardStatusRequest, redcardId: Long) {
        redcardFacade.updateRedcardStatus(request, redcardId)
    }

    // TODO: auth 쪽 완성 시 테스트 진행
    fun checkUser(request: CheckUserRequest): CheckUserResponse {
        val user: User = userRepository.findByEmail(request.email) ?: throw UserNotFoundException()

        return CheckUserResponse(user.id!!, user.name)
    }

    fun getDonations(
        page: Int, size: Int
    ): PageResponse<DonationListResponse> {
        return donationFacade.getDonations(page, size)
    }
}